client/src/components/PopoverOptionSet/PopoverOptionSet.js

```diff
@@ -0,0 +1,213 @@
+          id="popover-option-set__search-input"
+          onChange={this.handleSearchValueChange}
+          placeholder={searchPlaceholder}
+          type="text"
+          value={searchValue}
+        />
+        {this.renderSearchValueClearLink()}
+      </InputGroup>
+    );
+  }
+
+  /**
+   * Render either all buttons available, buttons matching the search term, or a message that
+   * there are no matching buttons
+   * @returns {DOMElement}
+   */
+  renderOptionButtons() {
+    const { onSearch, onButtonClick, ButtonComponent } = this.props;
+    let { buttons } = this.props;
+    const { searchValue } = this.state;
+
+    if (searchValue.length !== 0) {
+      buttons = onSearch(searchValue, buttons);
+    }
+
+    if (buttons.length === 0) {
+      return (
+        <div className="popover-option-set__button-container">
+          <div className="popover-option-set__no-results">
+            {i18n._t('PopoverOptionSet.NO_RESULTS', 'No results found')}
+          </div>
+        </div>
+      );
+    }
+
+    return (
+      <div className="popover-option-set__button-container">
+        {buttons.map((button) => (
+          <ButtonComponent
+            className={
+              classNames(
+                button.className,
+                'popover-option-set__button'
+              )
+            }
+            key={button.key}
+            onClick={onButtonClick(button)}
+          >
+            {button.content}
+          </ButtonComponent>
+        ))}
+      </div>
+    );
+  }
+
+  /**
+   * Render the option set popover
+   * @returns {DOMElement}
+   */
+  render() {
+    const { container, extraClass, isOpen, placement, target } = this.props;
+    const popoverClassNames = classNames(
+      'popover-option-set',
+      extraClass
+    );
+
+    return (
+      <Popover
+        className={popoverClassNames}
+        container={container}
+        hideArrow
+        isOpen={isOpen}
+        placement={placement}
+        target={target}
+        toggle={this.handleToggle}
+      >
+        {this.renderSearchBox()}
+        {this.renderOptionButtons()}
+      </Popover>
+    );
+  }
+}
+
+const buttonType = PropTypes.shape({
+  key: PropTypes.string.required,
+  content: PropTypes.node.required,
+  className: PropTypes.oneOfType([
+    PropTypes.string,
+    PropTypes.object,
+    PropTypes.arrayOf(PropTypes.string)
+  ]),
+});
+
+PopoverOptionSet.propTypes = {
+  buttons: PropTypes.arrayOf(buttonType),
+  // Accepts a function that takes a search term as a first parameter and a set
+  // of buttons to match against that returns a filtered set of buttons
+  // Default search handler assumes button content to be plain text and performs
+  // a simple string.contains check.
+  onSearch: PropTypes.func,
+  // function that accepts a button object & returns an event handler
+  // e.g. (button) => (event) => event.preventDefault() && togglePopover();
+  onButtonClick: PropTypes.func.isRequired,
```

Comment:
I'm all ears for a better name for this - or perhaps a more accepted way of doing something like this?

Reply:
Right - if it's not an event handler itself then we shouldn't call it `onXYZClick`. Maybe something like `provideButtonClickHandler` or something?